posts/2015-07-25-new-javascript-generators.md

```diff
@@ -0,0 +1,117 @@
+---
+title: New JavaScript Generators
+author: Freezeboy
+date: 2015-07-25 00:00
+toc: true
+---
+
+# Summary
+
+In the upcoming Servant version, it will be way easier to build a javascript
+connector to your webservice. The chosen API is *as usual with Servant*
+putting the hard work away from your eyes to let you concentrate on what you are
+working on in most situations.
+
+# The Problem
+
+Until recently, we had a javascript generator using exclusively
+JQuery, but depending on the framework you might want to use in your project,
+it can be overkill to fetch [JQuery](http://jquery.org), or else you are using
+a library that prefers doing things differently, such as 
+[AngularJS](http://angular.io), which contains its own solution to handle
+Ajax calls. On the other hand, sometimes your need is so simple that you don't even want JQuery.
+
+One solution could have been to have multiple packages `servant-jquery`,
+`servant-angular`, and so on... But we chose to put them all in the same
+package because the generators share lots of *problems & solutions*.
+
+We have to provide a function for each endpoint defined in the API, with as many
+parameter as needed. In most cases the client application will simply call
+these functions throughout its flow.
+
+# How do I use it?
+
+A complete example is present in the [tutorial](/tutorial/javascript.html).
+
+Basically you can define your API as [usual](/tutorial/api-type.html),
+``` haskell
+data ServerAPI = realAPI    -- ^ This is the real api 
+		 :<|> Raw   -- ^ As a helper, we will provide the
+			    -- js file at the root of the API
+
+serverApi :: Proxy ServerAPI
+serverApi = Proxy
+```
+
+And now, we just have to generate the client code
+``` haskell
+generateApi :: IO ()
+generateApi = do
+	-- we generate different versions of the API
+	writeJsForApi realAPI angular "angular.js"
+	writeJsForApi realAPI vanilla "vanilla.js"
+```
+
+And we [serve](/tutorial/server.html) the bundle to the world
+``` haskell
+main :: IO ()
+main = do
+	generateApi
+	-- And now let serve it
+	run (serve serverApi) 8000
+```
+
+# What changed
+
+## API simplification
+
+Since `0.4` branch, we drastically simplified the API so that it is easier
+to use and to extend (if you want to use the *next new javascript framework*.
+
+## More options
+
+Yes it may sound counter-intuitive with previous paragraph... but It's not!
+
+A simple "configuration" mecanism lets you
+
+- Rename *success* and *error* callbacks
+- prefix your function definitions (to put them in a namespace, or simply
+change the name)
+- even customize the naming heuristics
+
+Some generators have even more options (Namely Angular).
```

Comment:
Then it's just one generator =)

So something like "The angular generator even has options specific to Angular.js, unlike the jQuery and vanilla Javascript ones"